File: question/leetcode/leetcode0102.Binary-Tree-Level-Order-Traversal/leetcode0102.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode() {
	}

	TreeNode(int val) {
		this.val = val;
	}

	TreeNode(int val, TreeNode left, TreeNode right) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

class Solution {
	public List<List<Integer>> levelOrder(TreeNode root) {
		if (root == null) {
			return new ArrayList<>();
		}
		List<List<Integer>> ans = new ArrayList<>();
		Queue<TreeNode> queue = new LinkedList<>();
		queue.offer(root);

		while (!queue.isEmpty()) {
			int size = queue.size();
			List<Integer> levelValues = new ArrayList<>();

			for (int i = 0; i < size; i++) {
				TreeNode currNode = queue.poll();
				levelValues.add(currNode.val);

				if (currNode.left != null) {
					queue.offer(currNode.left);
				}
				if (currNode.right != null) {
					queue.offer(currNode.right);
				}
			}

			ans.add(levelValues);
		}
		return ans;
	}
}

File: question/leetcode/leetcode0102.Binary-Tree-Level-Order-Traversal/leetcode0102.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = (val === undefined ? 0 : val)
		this.left = (left === undefined ? null : left)
		this.right = (right === undefined ? null : right)
	}
}

function levelOrder(root: TreeNode | null): number[][] {
	if (!root) { return [] };

	const ans: number[][] = [];
	let queue: TreeNode[] = [root];

	while (queue.length > 0) {
		const levelValues: number[] = []
		const nextQueue: TreeNode[] = [];

		for (const node of queue) {
			levelValues.push(node.val);
			if (node.left) nextQueue.push(node.left);
			if (node.right) nextQueue.push(node.right);
		}

		ans.push(levelValues);
		queue = nextQueue;
	}

	return ans;
};

function levelOrderII(root: TreeNode | null): number[][] {
	if (!root) { return [] };

	const ans: number[][] = [];
	let queue: TreeNode[] = [root];

	while (queue.length > 0) {
		const size = queue.length;
		const levelValues: number[] = [];
		for (let i = 0; i < size; i++) {
			const currentNode = queue.shift();
			levelValues.push(currentNode!.val)
			if (currentNode!.left) {
				queue.push(currentNode!.left);
			}
			if (currentNode!.right) {
				queue.push(currentNode!.right);
			}
		}
		ans.push(levelValues);
	}

	return ans;
};
